zcash_primitives/src/transaction/components/sapling.rs

```diff
@@ -1,36 +1,107 @@
+use core::fmt::Debug;
 use ff::PrimeField;
 use group::GroupEncoding;
-
 use std::io::{self, Read, Write};
 
-use zcash_note_encryption::ShieldedOutput;
+use zcash_note_encryption::{ShieldedOutput, COMPACT_NOTE_SIZE};
 
 use crate::{
     consensus,
     sapling::{
         note_encryption::SaplingDomain,
-        redjubjub::{PublicKey, Signature},
+        redjubjub::{self, PublicKey, Signature},
         Nullifier,
     },
 };
 
-use zcash_note_encryption::COMPACT_NOTE_SIZE;
+use super::{amount::Amount, GROTH_PROOF_SIZE};
 
-use super::GROTH_PROOF_SIZE;
+pub type GrothProofBytes = [u8; GROTH_PROOF_SIZE];
 
 pub mod builder;
 
+pub trait Authorization: Debug {
+    type Proof: Clone + Debug;
+    type Signature: Clone + Debug;
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct Unproven {}
+
+impl Authorization for Unproven {
+    type Proof = ();
+    type Signature = ();
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct Unauthorized {}
+
+impl Authorization for Unauthorized {
+    type Proof = GrothProofBytes;
+    type Signature = ();
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct Authorized {}
+
+impl Authorization for Authorized {
+    type Proof = GrothProofBytes;
+    type Signature = redjubjub::Signature;
```

Comment:
This should be two different types (in preparation for when we move to the `redjubjub` crate), for spendAuth and binding.